<p><img alt="Colaboratory logo" width="100" height="130" src="https://www.udea.edu.co/wps/wcm/connect/udea/99fc43e7-7a64-45bd-97fc-96639b70813d/logosimbolo-vertical.png?MOD=AJPERES&CVID=ljeLvHr" align="left" hspace="10px" ></p>

<h1><b>Tarea 2: Física experimental <b></h1>

----
<div align="right">
<font size=3>
José David Bernal Restrepo <br>
Instituto de física
</font>
</div>

Esta actividad tiene como objetivo dominar los principios de la teoría elemental de propagación de la incertidumbre
deduciendo y calculando la incertidumbre experimental en una medida a partir de las reglas correctas y siguiendo los
principios de la lógica experimental.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


## 1. Medidas de longitud
En el presente estudio, se abordará la descripción de la situación física que concierne a un péndulo simple con una longitud fija de $(88.3 \pm 0.1)cm$. En esta investigación, no se contempla la toma de múltiples medidas de la longitud del péndulo, dado que este parámetro permanece constante a lo largo de todo el experimento y no es propenso a errores sistemáticos. Por lo tanto, una medida precisa de la longitud inicial es suficiente para mantener constante este valor a lo largo de todas las fases del estudio.

## 2. Medida de periodo de oscilación 
-**Es preferible que mida el tiempo de varios periodos y luego divida, ¿por qué?**

En el proceso de determinación del período de oscilación del péndulo, hemos contemplado la posibilidad de medir tanto el tiempo requerido para múltiples oscilaciones como el tiempo para una sola oscilación. Esta elección se fundamenta en la consideración de que, al registrar el tiempo de varias oscilaciones, podemos mitigar la incertidumbre derivada de errores sistemáticos, como el tiempo de reacción al detener el cronómetro.

En este contexto, consideramos la expresión que modela el período del péndulo (T) como sigue:  $$T=2\pi\cdot\sqrt{\frac{l}{g}}$$

<h5>carga de datos experimentales</h5>

In [2]:
# -- Importar datos --
data = pd.read_csv(r'../data/processed/Datos_pendulo.csv')

data['T(s)'] = data['Tiempo (s)']/data['Oscilaciones'] # get T(s) for every messuarement

data.sample(5)

Oscilaciones  Tiempo (s)   T(s)
24             5        9.41  1.882
7              1        1.93  1.930
19             5        9.56  1.912
12             1        1.77  1.770
20             5        9.65  1.930

### 2.1 Reporte de T

- **¿Qué estrategia debería usar para disminuir el error provocado por la manipulación humana del cronómetro?**


Como se ha destacado previamente, para aumentar la precisión en la recopilación de datos relacionados con el período de oscilación de un péndulo, se ha seleccionado la estrategia de medir el período considerando múltiples oscilaciones. Esta elección se basa en la premisa de que al registrar el período de varias oscilaciones en lugar de una sola, se reducen los efectos de errores sistemáticos que pueden surgir, como el tiempo de reacción al detener el cronómetro. Al medir múltiples oscilaciones y promediar los resultados, se logra un conjunto de datos más robusto y confiable, lo que a su vez mejora la precisión en la determinación del período del péndulo.

En aras de minimizar los errores potenciales relacionados con el tiempo de reacción, hemos implementado la práctica de que la misma persona que libera el péndulo sea la encargada de cronometrar su período de oscilación. Esta medida se ha adoptado con el propósito de asegurar una mayor consistencia y precisión en la toma de datos.
La lógica subyacente detrás de esta elección es que la persona que suelta el péndulo tiene una percepción inmediata del instante en que ocurre la liberación, lo que reduce el tiempo de reacción al iniciar la medición del período. Al minimizar este tiempo de reacción, podemos mitigar la introducción de errores sistemáticos en nuestros resultados, lo que, en última instancia, contribuye a la obtención de mediciones más confiables y precisas del período de oscilación del péndulo.

Reportaremos el periodo cómo: $$T = \bar{t} \pm S_x$$

de dónde $S_x$ es la suma en cuadratura de la incertidumbre estándar y la resolución de ilnstrumento $S_x = \sqrt{ (\frac{s}{\sqrt{N}})^2 + (\delta x)^2}$

In [29]:
#  Medidas de tendencia central

t_array = data['T(s)'].values  #Arreglo con los T(s)
std = np.std(t_array)
mean = np.mean(t_array)

# Incertidumbre estandar
error_estandar = std/len(t_array)

# Resolución del instrumento
error_cronometro = 1/100 # (s)

# Error total
error_total = np.sqrt(error_cronometro**2 + error_estandar**2)

print(f'T = ({mean :.2f} ± {error_total :.2f}) s')


T = (1.88 ± 0.01) s


El resultado de $T = (1.88 \pm 0.01) s$ es un informe adecuado si solo consideramos la incertidumbre estándar y la incertidumbre instrumental. Sin embargo, es importante tener en cuenta que esta representación no incluye errores sistemáticos en la medición del tiempo. Cuando consideramos estos errores sistemáticos, debemos aumentar el valor de la incertidumbre para reflejar una estimación más completa y precisa de la medida.

Por lo tanto, al tener en cuenta los errores sistemáticos en la medición del tiempo, podríamos informar el resultado como $T = (1.88 \pm 0.02) s$, donde la incertidumbre se ha incrementado para reflejar la contribución de los errores sistemáticos en la medición cronometrada del tiempo. Esto proporciona una descripción más precisa y completa de la medida.

In [4]:
# Tabla de medidas estadisitcas para intervalos de datos
def stats(muestras):
    """Realiza cálculos estadísticos para un conjunto de datos de muestra
    Input:
        muestras (list):  Lista de tamaños de muestra a considerar.
        
    Output: 
        pd.DataFrame:  DataFrame con medidas estadísticas evaluadas con cada tamaño de muestra."""

    error_instrumento = 1/100
    resultados = []

    for muestra in muestras: 

        muestra_datos = t_array[:muestra]

        #DevS
        std = np.std(muestra_datos)

        # Incertidumbre estandar cuadrada
        error_est_cuadrado = (std/len(t_array))**2 

        # Incertidumbre mitad instrumento
        error_mitad_instrumento = np.round((error_instrumento / 2) ** 2, 5)

        # Incertidumbre total 
        error_total_muestra = np.round(np.sqrt(error_est_cuadrado + (error_instrumento ** 2)), 3)

        resultados.append({
            'Tamaño de Muestra': muestra, 
            'Media':  np.round(np.mean(muestra_datos),3),
            'Desviación Estándar': np.round(std,3),
            'Incertidumbre Estándar Cuadrado': error_est_cuadrado,
            'Incertidumbre Mitad Instrumento Cuadrado': error_mitad_instrumento,
            'Incertidumbre Total': error_total_muestra
        })

    return pd.DataFrame(resultados)

stats([5, 10, 15, 20, 25, 30])

Tamaño de Muestra  Media  Desviación Estándar  \
0                  5  1.828                0.083   
1                 10  1.853                0.078   
2                 15  1.851                0.075   
3                 20  1.870                0.073   
4                 25  1.876                0.067   
5                 30  1.885                0.065   

   Incertidumbre Estándar Cuadrado  Incertidumbre Mitad Instrumento Cuadrado  \
0                         0.000008                                   0.00002   
1                         0.000007                                   0.00002   
2                         0.000006                                   0.00002   
3                         0.000006                                   0.00002   
4                         0.000005                                   0.00002   
5                         0.000005                                   0.00002   

   Incertidumbre Total  
0                 0.01  
1                 0.01  
2                 0.01  
3                 0.01  
4                 0.01  
5                 0.01

En este análisis, podemos observar una tendencia consistente en el comportamiento de la media a medida que aumenta el tamaño de nuestro conjunto de datos. Además, es notable que la dispersión de los datos, representada por la desviación estándar, tiende a disminuir a medida que el grupo de datos se amplía. Esta reducción es aún más evidente después de los primeros 15 resultados, cuando se consideran múltiples oscilaciones.

In [30]:
g = lambda l, T: (l*4*np.pi**2)/(T**2) # g in funtion of l and T
print(f'{g(0.883, 1.88) :.2f}')


9.86


## 3. Determinación de g


Utilizando la relación matemática $T = 2\pi\sqrt{\frac{l}{g}}$, junto con nuestras medidas directas de $l$ y $T$, podemos derivar una expresión que nos permite calcular la aceleración de la gravedad, $g$, de la siguiente manera: 
$$g =\frac{4\pi^2\cdot l}{T^2}$$
Al sustituir los valores obtenidos en nuestros datos experimentales en esta ecuación, obtenemos un valor aproximado de $9.86 \frac{m}{s^2}$ para la aceleración de la gravedad.

Para determinar la incertidumbre en nuestras mediciones, emplearemos dos métodos distintos: la Teoría Elemental de la Incertidumbre, considerando los valores extremos, y el Método de Propagación de Errores mediante derivadas parciales.

1. Estime las incertidumbres relativas por ambos métodos, compárelas y discuta.
<h5>Teoria elemental de la incertidumbre </h5>

$$g_{min} = 9.45 \frac{m}{s^2}$$
$$g_{max} = 10.30  \frac{m}{s^2}$$
Para asegurar que la diferencia sea simétrica, podemos expresar la aceleración debida a la gravedad como:

$$g = (9.9 \pm 0.4) \frac{m}{s^2}$$

De esta manera, estamos representando la incertidumbre de manera simétrica alrededor del valor central, lo que proporciona una descripción más precisa de la aceleración debida a la gravedad en nuestro informe.

<h5>Derivadas parciales </h5>
Otro método para calcular la incertidumbre de una variable de medida indirecta es utilizar una aproximación para incertidumbres pequeñas a través del cálculo de derivadas parciales.

$$\Delta g = \frac{\partial g}{\partial l}\Big |_l \cdot \Delta l + \frac{\partial g}{\partial T} \Big |_T \cdot \Delta T$$

así: $$\Delta g = \frac{4\pi^2}{(1.88 s)^2}\cdot(0.001 m) - \frac{2\pi^2(0.853 m)}{(1.88 s)^3} \cdot(0.02s) = 0.04 \frac{m}{s^2}$$

$$g = (9.86 \pm 0.04) \frac{m}{s^2}$$

que significativamente es un error con un orden de magnitud diferente al obtenido por la teoria elemental de la incertidumbre, en mi opinión el uso del método de derivadas parciales es mucho más efectivo ya que usar los valores exremos estamos incluyendo outliers que poco benefician en el reporte de la incertidumbre

2. ¿Podría afirmar que una de las variables presenta una incertidumbre dominante respecto a la otra en el contexto del análisis y propagación de la incertidumbre?

La incertidumbre ocasionada por el tiempo es considerablemente mayor a la ocasionada por la logitud ya que la expresión $ \frac{\partial g}{\partial T} \Big |_T \cdot \Delta T$ es considerablemente mayor 

3.  La siguiente expresión da una buena estimación del valor de la aceleración gravitacional terrestre en función de la latitud geográfica (l) y la altura en metros sobre el nivel del mar (h):

$$g\ = 9.780327 (1A(l)- B(2l))-(3.086\times{10}^{-6})h\frac{m}{s^2}$$
donde A = 0.0053024 y B =0.0000058.
Tome lo anterior y úselo para comparar sus resultados con el valor esperado para la aceleración gravitacional en el laboratorio. 

usando $l = 6.2°$ (en radianes ) y $h = 1500 m $, obtenemos $g_{\text esperado} = 9.78$, así nos es posible definir un error relativo porcentual
 $$\Delta X_\% = \frac{|X_{esperado} - X_{medido}|}{X_{esperado}} 100\%$$
 $$\Delta g_\% = \frac{|9.78 \frac{m}{s^2}-9.86 \frac{m}{s^2}|}{9.7 \frac{m}{s^2}} 100\% \approx 0.8\% $$



